### PR TITLE
[WIP] GetObject API: Return ReadCloser & ObjectInfo{}

### PR DESCRIPTION
**Pull request by vadmeste**
Still work in progress, and possibly needs an additional change in GetObject() API.

Not well tested yet but go test is successful.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: cmd/fs-v1.go [hunk 1]
```diff
@@ -375,91 +366,63 @@ func (fs fsObjects) CopyObject(srcBucket, srcObject, dstBucket, dstObject string
...
+	if err != nil {
+		return nil, ObjectInfo{}, toObjectErr(err, bucket, object)
 	}
 
-	// Reply back invalid range if the input offset and length fall out of range.
-	if offset > size || offset+length > size {
-		return traceError(InvalidRange{offset, length, size})
+	objReader = reader
+	if length >= 0 {
+		objReader = ioutil.NopCloser(io.LimitReader(reader, length))
```
Comment (harshavardhana) [inline cmd/fs-v1.go:418]: Don't use `NopCloser` it wouldn't close the underlying Close() on the 'reader' leading to fd leak.

File: cmd/xl-v1-object.go [hunk 2]
```diff
@@ -186,36 +174,28 @@ func (xl xlObjects) GetObject(bucket, object string, startOffset int64, length i
...
-		return traceError(InvalidRange{startOffset, length, xlMeta.Stat.Size})
+		return nil, ObjectInfo{}, traceError(InvalidRange{offset, length, xlMeta.Stat.Size})
 	}
 
 	// Save the writer.
-	mw := writer
+
+	pipeReader, pipeWriter := io.Pipe()
+
+	mw := io.Writer(pipeWriter)
```
Comment (harshavardhana) [inline cmd/xl-v1-object.go:198]: Do you need to do this?

**Comment by harshavardhana** [comment]
You might have to rebase this PR @vadmeste

**Comment by harshavardhana** [comment]
what is going on with this PR @vadmeste ?

**Comment by vadmeste** [comment]
> what is going on with this PR @vadmeste ?

This is on hold since we are going to modify Object API interface to satisfy our gateway needs, won't we ?

**Comment by harshavardhana** [comment]
@vadmeste can we close this PR and take a look at this later?

**Pull request closed without merging**